preserve module signature with multiple calls

avikchaudhuri opened the pull request:
Previously we would error when trying to preserve the call signature for a module when it was called multiple times. This PR can now do this without erroring. The fix is to propagate call indices in a few more places.

Note that while this works in the presence of params, buffers, and tensor constants, preserving call signatures for multiple calls to a module when buffers are mutated is not supported yet. This is future work. The main problem is that we do not have enough metadata to `copy_` mutated buffers at the end of each call to a module, so the next call can read those buffers at the beginning. Making this work will likely need some explicit tracking of intermediate values of mutated buffers when collecting metadata during functionalization in export.

Note also that we stop short of creating a single graph out of multiple graphs: that is still future work. So the unflattened module will still have different targets `n`, `n@1`, `n@2`, etc. for each call when we ask the module call signature of `n` to be preserved. However it is way easier to swap all of these targets with a replacement that behaves similar to the original, because all of these calls will respect the original module call signature. (In particular, any constant inputs will be carried by the calls.)


Differential Revision: D64406945

avikchaudhuri commented on torch/export/unflatten.py:
```diff
@@ -779,16 +782,7 @@ def __init__(
         self.parent_call_module: Optional[torch.fx.Node] = None
         if parent is not None:
-            if self.fqn in module_call_graph and num_calls == 1:
-                raise ValueError(
```
main thing than changed

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -6140,24 +6140,26 @@ def forward(self, x):
+        if not is_retracebility_test(self._testMethodName):
+            test(
+                export(M(), inp, preserve_module_call_signature=("n",)),
+                swap={"n": N(), "n@1": N()},
```
easy swap

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -6140,24 +6140,26 @@ def forward(self, x):
-                "Cannot unflatten multiple calls to module n while preserving its signature",
-            ):
-                torch.export.unflatten(ep)
+        def test(ep, swap):
```
Can you move this test to test_unfatten.py? I want to make sure it works with training IR as well. This here tests training IR + run_decomp so it is still operating on a functional IR.

avikchaudhuri replied:
I don't like `test_unflatten.py` because in general there are more variations tested here. But good point that I should test this with state as well. I'll hold off landing until I do that / add them in.

For training IR only I could do a separate explicit test?

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -505,25 +505,29 @@ def _get_module_hierarchy(mod: torch.nn.Module) -> Dict[str, str]:
+    additional = [
+        ModuleCallEntry(fqn=fqn, signature=signature)
+        for fqn, signature in module_call_signatures.items()
+        if fqn not in _EXPORT_MODULE_HIERARCHY  # type: ignore[operator]
```
Is this because we have entries like @ now?

avikchaudhuri replied:
Yes.

tugsbayasgalan commented on torch/export/unflatten.py:
```diff
@@ -779,16 +782,7 @@ def __init__(
-            # generate call name for self.fqn
-            child_fqn = _call_name(self.fqn, num_calls + 1)
-            accessor = _compute_accessor(parent.fqn, child_fqn)
+            accessor = _compute_accessor(parent.fqn, self.child_fqn)
```
Does the accesor now can have @ in their name?

avikchaudhuri replied:
Accessor already potentially had `@` in the name, just moved `child_fqn` to an attribute so that it can be accessed later in `self.finalize_outputs(...)` without recomputing.

avikchaudhuri committed:
preserve module signature multiple calls (#137999)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/137999

Test Plan: fixed tests

Reviewed By: tugsbayasgalan

Differential Revision: D64406945

avikchaudhuri commented:
@pytorchbot merge -f "Landed internally"
